File: src/main/java/com/kata/bank_account/domain/shared/model/SavingAccount.java
package com.kata.bank_account.domain.shared.model;

import com.kata.bank_account.domain.shared.constant.AccountType;
import com.kata.bank_account.domain.shared.exception.AboveCeilingException;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

public class SavingAccount extends BankAccount {

    private static final AccountType accountType = AccountType.SAVING;
    private final BigDecimal ceiling;

    public SavingAccount(BigDecimal ceiling) {
        this.ceiling = ceiling;
    }

    public BigDecimal ceiling() {
        return ceiling;
    }

    @Override
    public void deposit(BigDecimal amount) {
        if (amount.compareTo(ceiling) >= 0) {
            throw new AboveCeilingException();
        }
        super.deposit(amount);
    }

    @Override
    public AccountType type() {
        return accountType;
    }
}

File: src/main/kotlin/com/kata/bank/domain/model/SavingAccount.kt
package com.kata.bank.domain.model

import com.kata.bank.domain.exception.AboveCeilingException

open class SavingAccount(private val ceiling: Long) : BankAccount() {

    override fun deposit(amount: Long) {
        if (aboveCeilingOnDeposit(amount)) {
            throw AboveCeilingException()
        }
        super.deposit(amount)
    }

    private fun aboveCeilingOnDeposit(amount: Long) = (balance() + amount) > ceiling
    override fun type(): AccountType {
        return AccountType.SAVING
    }
}
